ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SlidingWindowRateLimiter.java

```diff
@@ -0,0 +1,116 @@
+ * debt. This is due to not knowing the size of the response upfront.
+ */
+
+public class SlidingWindowRateLimiter {
+
+  private static long NUM_BYTES_IN_ONE_MEGABYTE = 1 * 1024 * 1024;
+
+  /**
+   * The log of all the responses returned in the past hour.
+   * It is a Queue ofPairs of (timestamp in milliseconds, response size in Bytes).
+   */
+  private final Queue<Pair<Long, Long>> responseSizesLog;
+
+  /**
+   * Throttle limit measured in Bytes.
+   */
+  private final long throttleLimit;
+
+  /**
+   * Window size over which the throttle is supposed to be enforced measured in milliseconds.
+   */
+  private final long slidingWindowSizeMs;
+
+  /**
+   * Aggregate of pull query response sizes in the past hour
+   */
+  private long numBytesInWindow;
+
+  public SlidingWindowRateLimiter(final int requestLimitInMB, final long slidingWindowSizeMs) {
+    checkArgument(requestLimitInMB >= 0,
+            "Pull Query bandwidth limit can't be negative.");
+    checkArgument(slidingWindowSizeMs >= 0,
+            "Pull Query throttle window size can't be negative");
+
+    this.throttleLimit = (long) requestLimitInMB * NUM_BYTES_IN_ONE_MEGABYTE;
+    this.slidingWindowSizeMs = slidingWindowSizeMs;
+    this.responseSizesLog = new LinkedList<>();
+    this.numBytesInWindow = 0;
+  }
+
+  /**
+   * Checks if pull queries have returned more than the throttleLimit in the past hour.
+   * Throws a KsqlException is the limit has been breached
+   * @throws KsqlException Exception that the throttle limit has been reached for pull queries
+   */
+  public synchronized void allow() throws KsqlException {
+    this.allow(Time.SYSTEM.milliseconds());
+  }
+
+  @VisibleForTesting
+  protected synchronized void allow(final long timestamp) throws KsqlException {
+    checkArgument(timestamp >= 0,
+            "Timestamp can't be negative.");
+
+    while (!responseSizesLog.isEmpty()
+            && timestamp - responseSizesLog.peek().left >= slidingWindowSizeMs) {
+      this.numBytesInWindow -= responseSizesLog.poll().right;
+    }
+    if (this.numBytesInWindow > throttleLimit) {
+      throw new KsqlException("Host is at bandwidth rate limit for pull queries.");
+    }
+  }
+
+  /**
+   * Adds the responseSizeInBytes and its timestamp to the queue of all response sizes
+   * in the past hour.
+   * @param responseSizeInBytes pull query response size measured in Bytes
+   */
+  public synchronized void add(final long responseSizeInBytes) {
+    add(Time.SYSTEM.milliseconds(), responseSizeInBytes);
+  }
+
+  @VisibleForTesting
+  protected synchronized void add(final long timestamp, final long responseSizeInBytes) {
+    checkArgument(timestamp >= 0,
+            "Timestamp can't be negative.");
+    checkArgument(responseSizeInBytes >= 0,
+            "Response size can't be negative.");
+
+    responseSizesLog.add(new Pair<>(timestamp, responseSizeInBytes));
```

Comment:
Just wearing my paranoid window processing interviewer hat here :) Just in case we may have a time shift from `Time.SYSTEM.milliseconds()`, maybe do a quick sorted insert from the head than blinding adding at the head? In practice most of it would just end at the head so perf-wise it should be fine.

Reply:
Since the purpose of the RateLimiter was to bound the bandwidth usage approximately over the span of an hour (on that time-scale, I don't see a significant time shift happening), I think the current algorithm will work just fine. Do you think it is worth making this optimization?